dbms/src/Storages/DeltaMerge/Index/InvertedIndex/Reader.cpp

```diff
@@ -0,0 +1,248 @@
+    case TypeIndex::Int32:
+        return std::make_shared<InvertedIndexMemoryReader<Int32>>(buf, index_size);
+    case TypeIndex::UInt64:
+        return std::make_shared<InvertedIndexMemoryReader<UInt64>>(buf, index_size);
+    case TypeIndex::Int64:
+        return std::make_shared<InvertedIndexMemoryReader<Int64>>(buf, index_size);
+    case TypeIndex::Date:
+        return std::make_shared<InvertedIndexMemoryReader<UInt16>>(buf, index_size);
+    case TypeIndex::DateTime:
+        return std::make_shared<InvertedIndexMemoryReader<UInt32>>(buf, index_size);
+    case TypeIndex::Enum8:
+        return std::make_shared<InvertedIndexMemoryReader<Int8>>(buf, index_size);
+    case TypeIndex::Enum16:
+        return std::make_shared<InvertedIndexMemoryReader<Int16>>(buf, index_size);
+    case TypeIndex::MyDate:
+    case TypeIndex::MyDateTime:
+    case TypeIndex::MyTimeStamp:
+        return std::make_shared<InvertedIndexMemoryReader<UInt64>>(buf, index_size);
+    case TypeIndex::MyTime:
+        return std::make_shared<InvertedIndexMemoryReader<Int64>>(buf, index_size);
+    default:
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unsupported type_id: {}", magic_enum::enum_name(type_id));
+    }
+}
+
+template <typename T>
+void InvertedIndexMemoryReader<T>::load(ReadBuffer & read_buf, size_t index_size)
+{
+    // 1. read all data
+    std::vector<char> buf(index_size);
+    RUNTIME_CHECK(read_buf.readBig(buf.data(), index_size) == index_size);
+
+    // 2. check magic flag
+    size_t data_size = index_size - InvertedIndex::MagicFlagLength;
+    if (memcmp(buf.data() + data_size, InvertedIndex::MagicFlag, InvertedIndex::MagicFlagLength) != 0)
+        throw Exception(ErrorCodes::ABORTED, "Invalid magic flag");
+
+    // 3. read meta size
+    data_size = data_size - sizeof(UInt32);
+    UInt32 meta_size = *reinterpret_cast<const UInt32 *>(buf.data() + data_size);
+
+    // 4. read meta
+    ReadBufferFromMemory buffer(buf.data() + data_size - meta_size, meta_size);
+    InvertedIndex::Meta<T> meta;
+    data_size = data_size - meta_size;
+    InvertedIndex::Meta<T>::deserialize(meta, buffer);
+
+    // 5. read blocks & build index
+    buffer = ReadBufferFromMemory(buf.data(), data_size);
+    for (const auto meta_entry : meta.entries)
+    {
+        auto count = buffer.count();
+        InvertedIndex::Block<T> block;
+        InvertedIndex::Block<T>::deserialize(block, buffer);
+        RUNTIME_CHECK(buffer.count() - count == meta_entry.size);
+        for (const auto & block_entry : block.entries)
+        {
+            auto [value, row_ids] = block_entry;
+            index[value] = row_ids;
+        }
+    }
+}
+
+template <typename T>
+void InvertedIndexMemoryReader<T>::search(BitmapFilterPtr & bitmap_filter, const Key & key) const
+{
+    T real_key = key;
+    auto it = index.find(real_key);
+    if (it != index.end())
+        bitmap_filter->set(it->second, nullptr);
```

Comment:
Existing values in bitmap_filter is not cleared. Is it ok?

Reply:
ok